The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/26793196271/job/78983995489) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/78983995489) [(plain)](https://github.com/rust-lang/rust/commit/a43bb36d8c286207efff2d930ef65b9707c79b9a/checks/78983995489/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6878 files
tidy check
tidy [rustdoc_gui_tests (tests)]: /checkout/tests/rustdoc-gui/trait-impl-sort.goml: rustdoc-gui tests must start with a small description
tidy [rustdoc_gui_tests (tests)]: FAIL
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
---
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: rustdoc_gui_tests (tests)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1624:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1475:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1475:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1596:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1397:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1139:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1118:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:41
  local time: Tue Jun  2 01:50:04 UTC 2026
  network time: Tue, 02 Jun 2026 01:50:04 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/src/librustdoc/html/render/mod.rs
+++ b/src/librustdoc/html/render/mod.rs
@@ -79,6 +79,7 @@ use crate::html::format::{
 use crate::html::markdown::{
     HeadingOffset, IdMap, Markdown, MarkdownItemInfo, MarkdownSummaryLine, short_markdown_summary,
 };
+use crate::html::render::print_item::compare_names;
 use crate::html::render::search_index::get_function_type_for_search;
 use crate::html::static_files::SCRAPE_EXAMPLES_HELP_MD;
 use crate::html::{highlight, sources};
@@ -941,6 +942,16 @@ fn short_item_info(
     extra_info
 }
 
+// Prints the polarity and path of an impl's trait, if it has one, e.g. `Send`, `!Sync`.
+fn impl_trait_key(cx: &Context<'_>, i: &Impl) -> Option<String> {
+    let trait_ = i.inner_impl().trait_.as_ref()?;
+    let prefix = match i.inner_impl().polarity {
+        ty::ImplPolarity::Positive | ty::ImplPolarity::Reservation => "",
+        ty::ImplPolarity::Negative => "!",
+    };
+    Some(format!("{prefix}{:#}", print_path(trait_, cx)))
+}
+
 // Render the list of items inside one of the sections "Trait Implementations",
 // "Auto Trait Implementations," "Blanket Trait Implementations" (on struct/enum pages).
 fn render_impls(
@@ -950,7 +961,9 @@ fn render_impls(
     containing_item: &clean::Item,
     toggle_open_by_default: bool,
 ) -> fmt::Result {
-    let mut rendered_impls = impls
+    // Render each impl alongside its `impl_trait_key`, which is used as the primary sorting key
+    // to match the impl order in the sidebar.
+    let mut keyed_rendered_impls = impls
         .iter()
         .map(|i| {
             let did = i.trait_did().unwrap();
@@ -971,11 +984,16 @@ fn render_impls(
                     toggle_open_by_default,
                 },
             );
-            imp.to_string()
+            (impl_trait_key(cx, i).unwrap(), imp.to_string())
         })
         .collect::<Vec<_>>();
-    rendered_impls.sort();
-    w.write_str(&rendered_impls.join(""))
+
+    // Sort and then remove the `impl_trait_key`s, which are no longer needed after sorting.
+    keyed_rendered_impls
+        .sort_by(|(k1, h1), (k2, h2)| compare_names(k1, k2).then_with(|| h1.cmp(h2)));
+    let joined: String = keyed_rendered_impls.into_iter().map(|a| a.1).collect();
+
+    w.write_str(&joined)
 }
 
 /// Build a (possibly empty) `href` attribute (a key-value pair) for the given associated item.

--- a/src/librustdoc/html/render/sidebar.rs
+++ b/src/librustdoc/html/render/sidebar.rs
@@ -6,10 +6,10 @@ use askama::Template;
 use rustc_data_structures::fx::FxHashSet;
 use rustc_hir::def::CtorKind;
 use rustc_hir::def_id::{DefIdMap, DefIdSet};
-use rustc_middle::ty::{self, TyCtxt};
+use rustc_middle::ty::TyCtxt;
 use tracing::debug;
 
-use super::{Context, ItemSection, item_ty_to_section};
+use super::{Context, ItemSection, impl_trait_key, item_ty_to_section};
 use crate::clean;
 use crate::formats::Impl;
 use crate::formats::item_type::ItemType;
@@ -707,15 +707,9 @@ fn sidebar_render_assoc_items(
 
         let mut ret = impls
             .iter()
-            .filter_map(|it| {
-                let trait_ = it.inner_impl().trait_.as_ref()?;
-                let encoded = id_map.derive(super::get_id_for_impl(cx.tcx(), it.impl_item.item_id));
-
-                let prefix = match it.inner_impl().polarity {
-                    ty::ImplPolarity::Positive | ty::ImplPolarity::Reservation => "",
-                    ty::ImplPolarity::Negative => "!",
-                };
-                let generated = Link::new(encoded, format!("{prefix}{:#}", print_path(trait_, cx)));
+            .filter_map(|i| {
+                let encoded = id_map.derive(super::get_id_for_impl(cx.tcx(), i.impl_item.item_id));
+                let generated = Link::new(encoded, impl_trait_key(cx, i)?);
                 if links.insert(generated.clone()) { Some(generated) } else { None }
             })
             .collect::<Vec<Link<'static>>>();

--- a/tests/rustdoc-gui/search-tab.goml
+++ b/tests/rustdoc-gui/search-tab.goml
@@ -81,7 +81,7 @@ set-window-size: (851, 600)
 // Check the size and count in tabs
 assert-text: ("#search-tabs > button:nth-child(1) > .count", " (25) ")
 assert-text: ("#search-tabs > button:nth-child(2) > .count", " (7)  ")
-assert-text: ("#search-tabs > button:nth-child(3) > .count", " (0)  ")
+assert-text: ("#search-tabs > button:nth-child(3) > .count", " (1)  ")
 store-property: ("#search-tabs > button:nth-child(1)", {"offsetWidth": buttonWidth})
 assert-property: ("#search-tabs > button:nth-child(2)", {"offsetWidth": |buttonWidth|})
 assert-property: ("#search-tabs > button:nth-child(3)", {"offsetWidth": |buttonWidth|})

--- a/tests/rustdoc-gui/sidebar-foreign-impl-sort.goml
+++ b/tests/rustdoc-gui/sidebar-foreign-impl-sort.goml
@@ -1,4 +1,4 @@
-// Checks sidebar resizing close the Settings popover
+// Checks sidebar foreign impl ordering
 go-to: "file://" + |DOC_PATH| + "/test_docs/SidebarSort/trait.Sort.html#foreign-impls"
 
 // Check that the sidebar contains the expected foreign implementations

--- a/tests/rustdoc-gui/src/test_docs/lib.rs
+++ b/tests/rustdoc-gui/src/test_docs/lib.rs
@@ -10,6 +10,7 @@
 #![feature(associated_type_defaults)]
 #![feature(macro_attr)]
 #![feature(macro_derive)]
+#![feature(negative_impls)]
 
 /*!
 Enable the feature <span class="stab portability"><code>some-feature</code></span> to enjoy
@@ -87,6 +88,19 @@ impl AsRef<str> for Foo {
     fn as_ref(&self) -> &str {
         "hello"
     }
+}
+
+unsafe impl Send for Foo {}
+impl !Sync for Foo {}
+
+impl From<u8> for Foo {
+    fn from(value: u8) -> Self { todo!(); }
+}
+impl From<u16> for Foo {
+    fn from(value: u16) -> Self { todo!(); }
+}
+impl From<u32> for Foo {
+    fn from(value: u32) -> Self { todo!(); }
 }
 
 /// <div id="doc-warning-0" class="warning">I have warnings!</div>

--- a/tests/rustdoc-gui/trait-impl-sort.goml
+++ b/tests/rustdoc-gui/trait-impl-sort.goml
@@ -1,0 +1,20 @@
+// Check that trait impls in the sidebar and the main section have the same
+// ordering.
+
+go-to: "file://" + |DOC_PATH| + "/test_docs/struct.Foo.html"
+
+// .sidebar .trait-implementation li:nth-child(3) a
+assert-text: (".sidebar-elems .trait-implementation li:nth-child(1) a", "!Sync")
+assert-text: (".sidebar-elems .trait-implementation li:nth-child(2) a", "AsRef<str>")
+assert-text: (".sidebar-elems .trait-implementation li:nth-child(3) a", "From<u8>")
+assert-text: (".sidebar-elems .trait-implementation li:nth-child(4) a", "From<u16>")
+assert-text: (".sidebar-elems .trait-implementation li:nth-child(5) a", "From<u32>")
+assert-text: (".sidebar-elems .trait-implementation li:nth-child(6) a", "Send")
+
+// FIXME: this order doesn't match the sidebar order
+assert-text: ("#trait-implementations-list details:nth-child(1) .code-header", "impl AsRef<str> for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(2) .code-header", "impl From<u16> for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(3) .code-header", "impl From<u32> for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(4) .code-header", "impl From<u8> for Foo")
+assert-text: ("#trait-implementations-list section:nth-child(5) .code-header", "impl Send for Foo")
+assert-text: ("#trait-implementations-list section:nth-child(6) .code-header", "impl !Sync for Foo")

--- a/tests/rustdoc-gui/trait-impl-sort.goml
+++ b/tests/rustdoc-gui/trait-impl-sort.goml
@@ -11,10 +11,9 @@ assert-text: (".sidebar-elems .trait-implementation li:nth-child(4) a", "From<u1
 assert-text: (".sidebar-elems .trait-implementation li:nth-child(5) a", "From<u32>")
 assert-text: (".sidebar-elems .trait-implementation li:nth-child(6) a", "Send")
 
-// FIXME: this order doesn't match the sidebar order
-assert-text: ("#trait-implementations-list details:nth-child(1) .code-header", "impl AsRef<str> for Foo")
-assert-text: ("#trait-implementations-list details:nth-child(2) .code-header", "impl From<u16> for Foo")
-assert-text: ("#trait-implementations-list details:nth-child(3) .code-header", "impl From<u32> for Foo")
-assert-text: ("#trait-implementations-list details:nth-child(4) .code-header", "impl From<u8> for Foo")
-assert-text: ("#trait-implementations-list section:nth-child(5) .code-header", "impl Send for Foo")
-assert-text: ("#trait-implementations-list section:nth-child(6) .code-header", "impl !Sync for Foo")
+assert-text: ("#trait-implementations-list section:nth-child(1) .code-header", "impl !Sync for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(2) .code-header", "impl AsRef<str> for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(3) .code-header", "impl From<u8> for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(4) .code-header", "impl From<u16> for Foo")
+assert-text: ("#trait-implementations-list details:nth-child(5) .code-header", "impl From<u32> for Foo")
+assert-text: ("#trait-implementations-list section:nth-child(6) .code-header", "impl Send for Foo")

